lib/maths/analytics/CBoostedTreeFactory.cc

```diff
@@ -1135,138 +912,151 @@ void CBoostedTreeFactory::initializeUnsetEta(core::CDataFrame& frame) {
+    }
+    if (hyperparameters.eta().fixed() == false &&
+        hyperparameters.etaGrowthRatePerTree().rangeFixed() == false) {
+        double rate{m_TreeImpl->m_Hyperparameters.etaGrowthRatePerTree().value() - 1.0};
+        hyperparameters.etaGrowthRatePerTree().fixToRange(
+            1.0 + MIN_ETA_GROWTH_RATE_SCALE * rate, 1.0 + MAX_ETA_GROWTH_RATE_SCALE * rate);
     }
 }
 
 void CBoostedTreeFactory::initializeUnsetRetrainedTreeEta() {
-    if (m_TreeImpl->m_Hyperparameters.retrainedTreeEta().rangeFixed() == false) {
-        // The incremental loss function keeps the leaf weights around the
-        // magnitude of the old tree leaf weights so we search larger values
-        // of eta for trees we retrain.
-        m_TreeImpl->m_Hyperparameters.retrainedTreeEta().set(1.0);
-        m_LogRetrainedTreeEtaSearchInterval(MIN_PARAMETER_INDEX) =
-            common::CTools::stableLog(m_TreeImpl->m_Hyperparameters.eta().value());
-        m_LogRetrainedTreeEtaSearchInterval(MID_PARAMETER_INDEX) = 0.0;
-        m_LogRetrainedTreeEtaSearchInterval(MAX_PARAMETER_INDEX) = 0.0;
-
-        if (intervalIsEmpty(m_LogRetrainedTreeEtaSearchInterval)) {
-            m_TreeImpl->m_Hyperparameters.retrainedTreeEta().fix();
+
+    if (m_TreeImpl->m_Hyperparameters.incrementalTraining() == false) {
+        return;
+    }
+
+    skipIfAfter(CBoostedTreeImpl::E_NotInitialized, [&] {
+        if (m_TreeImpl->m_Hyperparameters.retrainedTreeEta().rangeFixed() == false) {
+            // The incremental loss function keeps the leaf weights around the
+            // magnitude of the old tree leaf weights so we search larger values
+            // of eta for trees we retrain.
+            auto& retrainedTreeEta = m_TreeImpl->m_Hyperparameters.retrainedTreeEta();
+            retrainedTreeEta.fixToRange(m_TreeImpl->m_Hyperparameters.eta().value(), 1.0);
+            retrainedTreeEta.set(1.0);
         }
-    } else {
-        m_LogRetrainedTreeEtaSearchInterval =
-            m_TreeImpl->m_Hyperparameters.retrainedTreeEta().fineTuneParameters();
+    });
+}
+
+void CBoostedTreeFactory::initializePredictionChangeCost() {
+
+    if (m_TreeImpl->m_Hyperparameters.incrementalTraining() == false) {
+        return;
     }
+
+    skipIfAfter(CBoostedTreeImpl::E_NotInitialized, [&] {
+        auto& hyperparameters = m_TreeImpl->m_Hyperparameters;
+        if (hyperparameters.predictionChangeCost().rangeFixed() == false) {
+            hyperparameters.predictionChangeCost().fixToRange(0.01, 2.0);
+            hyperparameters.predictionChangeCost().set(0.5);
+        }
+    });
 }
 
 void CBoostedTreeFactory::initializeUnsetTreeTopologyPenalty(core::CDataFrame& frame) {
 
-    auto& hyperparameters = m_TreeImpl->m_Hyperparameters;
+    if (m_TreeImpl->m_Hyperparameters.incrementalTraining() == false) {
+        return;
+    }
 
-    if (hyperparameters.treeTopologyChangePenalty().rangeFixed() == false) {
+    skipIfAfter(CBoostedTreeImpl::E_NotInitialized, [&] {
```

Comment:
Can we maybe get away without using these large unnamed lambda functions? I think it impairs the readability.

Reply:
I don't agree here. It is essentially no extra code other than just putting this inside a code block `{}`. This is the mechanism by which we ensure we reinitialise properly if we failover halfway through building and I think the mechanism is effective. Since it is really unrelated I think I will leave as is. Happy to discuss alternative approaches offline.